### PR TITLE
Make `get_current_state` cope with deleted labels

### PR DESCRIPTION
**Pull request by PeterJCLaw**
Previously it emitted an `IndexError` which was unhelpful and, in the one case which didn't already explicitly check for errors, somewhat silenced by us using `suppress_exceptions`.

Changes the return type from `State` to `Optional[State]`, updating usages to cope with this as needed. Also introduce specific tests to pin down the behaviour of `get_current_state`, including in the (unexpected) case where a label is in an invalid state.

### CODE REVIEW

File: routemaster/state_machine/tests/test_state_machine_api.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+from routemaster.state_machine import process_cron
+
+
+def test_handles_label_state_change_race_condition(app, create_deleted_label):
+    state_machine = app.config.state_machines['test_machine']
+    state = state_machine.states[1]
+
+    # Create a label which is not in the expected state. Doing this and then
+    # returning the affected label from the `get_labels` call is easier but
```
Comment (danpalmer) [inline routemaster/state_machine/tests/test_state_machine_api.py]: I'd replace "but" with "and" here, if I'm reading this correctly.

Comment (PeterJCLaw) [reply routemaster/state_machine/tests/test_state_machine_api.py]: ffa70055a42965db27dd8174b9ad53b5caba9b24

File: routemaster/state_machine/tests/test_state_machine_api.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
+from unittest import mock
```
[the remark this answers is not in the document]
Comment (PeterJCLaw) [reply routemaster/state_machine/tests/test_state_machine_api.py]: The tests in [`routemaster/state_machine/tests/test_state_machine.py`](https://github.com/thread/routemaster/blob/master/routemaster/state_machine/tests/test_state_machine.py) don't currently appear to test anything from the `routemaster.state_machine.api` module (based on the imports at least), hence the new file.

Would you still rather that this was in that file?

[the remark this answers is not in the document]
Comment (danpalmer) [reply routemaster/state_machine/tests/test_state_machine_api.py]: I think they do, they import the `state_machine` module and call the exported methods on it. I think everything in the api file is exported from the `state_machine` module?

**Comment by danpalmer** [comment]
@PeterJCLaw Can we have a test for the exception that's not currently covered?

File: routemaster/state_machine/tests/test_state_machine_api.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (PeterJCLaw) [reply routemaster/state_machine/tests/test_state_machine_api.py]: 579f117

**Comment by PeterJCLaw** [comment]
@danpalmer I believe 4ae739d covers the case you meant?

**Comment by danpalmer** [comment]
@PeterJCLaw Nice. Feel free to merge.

**Pull request merged**